Docs: Introduce Contributing Guide (#338)
grafana/grafana-plugin-examples: CONTRIBUTING.md

# Contributing

## How do I... <a name="toc"></a>

- [Use This Guide](#introduction)?
- Ask or Say Something? 🤔🐛😱
  - [Request Support](#request-support)
  - [Report an Error or Bug](#report-an-error-or-bug)
  - [Request a Feature](#request-a-feature)
- Make Something? 🤓👩🏽‍💻📜🍳
  - [Project Setup](#project-setup)
  - [Contributor License Agreement (CLA)](#contributor-license-agreement-cla)
  - [Contribute Documentation](#contribute-documentation)
  - [Contribute Code](#contribute-code)

## Introduction

> Thank you so much for your interest in contributing!. All types of contributions are encouraged and valued. See the [table of contents](#toc) for different ways to help and details about how this project handles them!📝

Reviewer: ```suggestion
Thank you so much for your interest in contributing! All types of contributions are encouraged and valued. See the [table of contents](#toc) for different ways to help and details about how this project handles them!📝
```

Please make sure to read the relevant section before making your contribution! It will make it a lot easier for us maintainers to make the most of it and smooth out the experience for all involved. 💚

The Project Team looks forward to your contributions. 🙌🏾✨

> ## Request Support

Reviewer 2: Is this section valid for examples?

Author: If something isn't working in an example how else could someone ask for assistance?

Reviewer 2: Bug report is below - for this report in particular I can't see what the difference would be?

## Report an Error or Bug

If you run into an error or bug with the project:

- Open an Issue at https://github.com/grafana/grafana-plugin-examples/issues
- Include _reproduction steps_ that someone else can follow to recreate the bug or error on their own.
- Provide project and platform versions (nodejs, npm, etc), depending on what seems relevant. If not, please be ready to provide that information if maintainers ask for it.

Once it's filed:

- The project team will label the issue.
- A team member will try to reproduce the issue with your provided steps. If there are no repro steps or no obvious way to reproduce the issue, the team will ask you for those steps. Bugs will not be addressed until they are reproduced.
- If the team is able to reproduce the issue it will be left to be [implemented by someone](#contribute-code).
- If you or the maintainers don't respond to an issue for 30 days, the issue will be closed. If you want to come back to it, reply (once, please), and we'll reopen the existing issue. Please avoid filing new issues as extensions of one you already made.

> ## Request a Feature

Reviewer 2: ```suggestion
## Request a new example or extension to an existing template
```

> If the project doesn't do something you need or want it to do:

Reviewer 2: ```suggestion
If our examples don't cover your use case and you would like to see a new template project available:
```

- Open an Issue at https://github.com/grafana/grafana-plugin-examples/issues

> - Provide as much context as you can about what you're running into.
> - Please try and be clear about why existing features and alternatives would not work for you.

Reviewer 2: ```suggestion
- Provide as much context as you can about what you are trying to do.
- Please indicate whether it should be incorporated within an existing template or a new example.
```

Note: The team is unlikely to be able to accept every single feature request that is filed. Please understand if they need to say no.

## Project Setup

So you wanna contribute some code! That's great! This project uses GitHub Pull Requests to manage contributions, so [read up on how to fork a GitHub project and file a PR](https://guides.github.com/activities/forking) if you've never done it before.

If this seems like a lot or you aren't able to do all this setup, you might also be able to [edit the files directly](https://help.github.com/articles/editing-files-in-another-user-s-repository/) without having to do any of this setup. Yes, [even code](#contribute-code).

If you want to go the usual route and run the project locally, though:

- [Install Node.js](https://nodejs.org/en/download/)
- [Fork the project](https://guides.github.com/activities/forking/#fork)

Then in your terminal:

- `cd path/to/your/clone/examples/<example_directory>`
- `npm install`
- `npm dev`
- `mage build:backend` (if the plugin has a backend)

> - `docker compose up`

Reviewer 2: ```suggestion
- `docker compose up --build`
```

Author: What's the reasoning for forcing builds? 🤔

Reviewer 2: It's what run server does

Author: Ok, let's keep it consistent as we discussed. 👍

And you should be ready to go!

> [!NOTE]
> This repository is a collection of Grafana plugins with their own dependency lock files, dev environments, etc. That is to say it is not a monorepo in the traditional sense. To develop against a plugin you must navigate inside it's directory and run the above commands for each example you would like to contribute to.

## Contributor License Agreement (CLA)

Before we can accept your pull request, you need to [sign our CLA](https://grafana.com/docs/grafana/latest/developers/cla/). If you haven't, our CLA assistant prompts you to when you create your pull request.

> ## Contribute Documentation

Reviewer 2: This section feels like overkill for readmes

Documentation contributions of any size are welcome! Feel free to file a PR even if you're just rewording a sentence to be more clear, or fixing a spelling mistake!

To contribute documentation:

- [Set up the project](#project-setup).
- Edit or add any relevant documentation.
- Make sure your changes are formatted correctly and consistently with the rest of the documentation.
- Re-read what you wrote, and run a spellchecker on it to make sure you didn't miss anything.
- Go to https://github.com/grafana/grafana-plugin-examples/pulls and open a new pull request with your changes.
- If your PR is connected to an open issue, add a line in your PR's description that says `Fixes: #123`, where `#123` is the number of the issue you're fixing.

Once you've filed the PR:

- One or more maintainers will use GitHub's review feature to review your PR.
- If the maintainer asks for any changes, edit your changes, push, and ask for another review.
- If the maintainer decides to pass on your PR, they will thank you for the contribution and explain why they won't be accepting the changes. That's ok! We still really appreciate you taking the time to do it, and we don't take that lightly. 💚

> - If your PR gets accepted, it will be marked as such, and merged into the `latest` branch soon after. Your contribution will be distributed to the masses once the PR is merged.

Reviewer 3: Should this be `main` branch?

```suggestion
- If your PR gets accepted, it will be marked as such, and merged into the `main` branch soon after. Your contribution will be distributed to the masses once the PR is merged.
```

## Contribute Code

> We like code commits a lot! They're super handy, and they keep the project going and doing the work it needs to do to be useful to others. Before considering contributing code please review [report an error or bug](#report-an-error-or-bug) and [request a feature](#request-a-feature) to make sure an issue has been filed and discussed with the project maintainers. PRs submitted without associated issues risk being closed or rejected.

Reviewer 3: The "request a feature" section doesn't exist. I think it should go to "Request a new example or extension to an existing template"

```suggestion
We like code commits a lot! They're super handy, and they keep the project going and doing the work it needs to do to be useful to others. Before considering contributing code please review [report an error or bug](#report-an-error-or-bug) and [request a new example](#request-a-new-example-or-extension-to-an-existing-template) to make sure an issue has been filed and discussed with the project maintainers. PRs submitted without associated issues risk being closed or rejected.
```

To contribute code:

- [Set up the project](#project-setup).
- Make any necessary changes to the source code.
- Include any [additional documentation](#contribute-documentation) the changes might need.
- Write tests that verify that your contribution works as expected.
- Go to https://github.com/grafana/grafana-plugin-examples/pulls and open a new pull request with your changes.
- If your PR is connected to an open issue, add a line in your PR's description that says `Fixes: #123`, where `#123` is the number of the issue you're fixing.

Once you've filed the PR:

- Barring special circumstances, maintainers will not review PRs until all checks pass (Travis, AppVeyor, etc).
- One or more maintainers will use GitHub's review feature to review your PR.
- If the maintainer asks for any changes, edit your changes, push, and ask for another review. Additional tags (such as `needs-tests`) will be added depending on the review.
- If the maintainer decides to pass on your PR, they will thank you for the contribution and explain why they won't be accepting the changes. That's ok! We still really appreciate you taking the time to do it, and we don't take that lightly. 💚

> - If your PR gets accepted, it will be marked as such, and merged into the `latest` branch soon after. Your contribution will be distributed to the masses next time the maintainers [tag a release](#tag-a-release)

Reviewer 3: ```suggestion
- If your PR gets accepted, it will be marked as such, and merged into the `main` branch soon after.
```
